File: samples/java/SupportChat/Server/Support/src/main/java/systems/zlink/samples/supportchat/server/support/actors/SupportUserActorRelocationAdapter.java
package systems.zlink.samples.supportchat.server.support.actors;

import com.fasterxml.jackson.databind.ObjectMapper;

import systems.zlink.framework.actors.ZLinkActorJoinOperationId;
import systems.zlink.framework.actors.ZLinkActorRelocationAdapter;
import systems.zlink.framework.actors.ZLinkRelocationCancellation;

import java.io.IOException;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class SupportUserActorRelocationAdapter
        implements ZLinkActorRelocationAdapter<SupportUserActor> {
    private static final ObjectMapper JSON = new ObjectMapper();

    @Override
    public CompletionStage<byte[]> capture(
            SupportUserActor actor, ZLinkRelocationCancellation cancellation) {
        try {
            return CompletableFuture.completedFuture(
                    JSON.writeValueAsBytes(
                            new TransferState(
                                    actor.displayName(),
                                    actor.role(),
                                    actor.participantId(),
                                    actor.conversationId(),
                                    actor.pendingConversationId(),
                                    actor.completedJoinOperations())));
        } catch (IOException error) {
            return CompletableFuture.failedFuture(error);
        }
    }

    @Override
    public CompletionStage<Void> restore(
            SupportUserActor actor, byte[] state, ZLinkRelocationCancellation cancellation) {
        try {
            TransferState transferred = JSON.readValue(state, TransferState.class);
            actor.setIdentity(
                    transferred.displayName(), transferred.role(), transferred.participantId());
            actor.joinConversation(transferred.conversationId());
            actor.restorePendingConversationJoin(transferred.pendingConversationId());
            actor.restoreCompletedJoinOperations(transferred.completedJoinOperations());
            return CompletableFuture.completedFuture(null);
        } catch (IOException error) {
            return CompletableFuture.failedFuture(error);
        }
    }

    public record TransferState(
            String displayName,
            String role,
            String participantId,
            String conversationId,
            String pendingConversationId,
            Set<ZLinkActorJoinOperationId> completedJoinOperations) {}
}

File: samples/kotlin/SupportChat/Server/Support/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/support/infrastructure/zlink/actors/SupportUserActorRelocationAdapter.kt
package systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.actors

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import systems.zlink.framework.actors.ZLinkActorJoinOperationId
import systems.zlink.framework.actors.ZLinkActorRelocationAdapter
import systems.zlink.framework.actors.ZLinkRelocationCancellation

class SupportUserActorRelocationAdapter : ZLinkActorRelocationAdapter<SupportUserActor> {
    override fun capture(
        actor: SupportUserActor,
        cancellation: ZLinkRelocationCancellation,
    ): CompletionStage<ByteArray> =
        CompletableFuture.completedFuture(
            json.writeValueAsBytes(
                TransferState(
                    actor.displayName,
                    actor.role,
                    actor.participantId,
                    actor.conversationId,
                    actor.pendingConversationId(),
                    actor.completedJoinOperations(),
                )
            )
        )

    override fun restore(
        actor: SupportUserActor,
        state: ByteArray,
        cancellation: ZLinkRelocationCancellation,
    ): CompletionStage<Void> {
        val transferred = json.readValue(state, TransferState::class.java)
        actor.setIdentity(transferred.displayName, transferred.role, transferred.participantId)
        if (transferred.conversationId.isNotBlank()) {
            actor.joinConversation(transferred.conversationId)
        }
        actor.restorePendingConversationJoin(transferred.pendingConversationId)
        actor.restoreCompletedJoinOperations(transferred.completedJoinOperations)
        return CompletableFuture.completedFuture(null)
    }

    companion object {
        private val json = jacksonObjectMapper()
    }

    data class TransferState(
        val displayName: String,
        val role: String,
        val participantId: String,
        val conversationId: String,
        val pendingConversationId: String,
        val completedJoinOperations: Set<ZLinkActorJoinOperationId>,
    )
}
